test/functional/api/v2/template_controller_test.rb

```diff
@@ -24,6 +25,146 @@ class TemplateControllerTest < ::ActionController::TestCase
+        assert_response :success
+        uptemplates = JSON.parse(@response.body)['message']['templates']
+        uptemplates.each do |template|
+          if template['name'] == 'filter1 template'
+            assert_not template['imported']
+            assert template['changed']
+          end
+        end
+      end
+
+      test "should force update locked template on import" do
+        post :import, params: { "repo" => "#{ForemanTemplates::Engine.root}/test/templates/core", "filter" => "filter1", "prefix" => '', "associate" => "new", "lock" => true }
+        File.open("#{ForemanTemplates::Engine.root}/test/templates/core/filter1.erb", "a") do |filt|
+          filt << "Appended from another test \n"
+        end
+        post :import, params: { "repo" => "#{ForemanTemplates::Engine.root}/test/templates/core", "filter" => "filter1", "prefix" => '', "associate" => "new", "lock" => true, "force" => true }
+        assert_response :success
+        updated = JSON.parse(@response.body)['message']['templates']
+        updated.each do |template|
+          if template['name'] == 'filter1 template'
+            puts template
+          end
+        end
+      end
+
+      test "should associate template to taxonomies in metadata on import" do
+        Dir.mktmpdir do |tmpdir|
+          t_name = "metadata_with_taxonomies"
+          org = taxonomies(:organization1)
+          loc = taxonomies(:location1)
+          create_local_template t_folder: tmpdir, t_filename: "metadata", t_name: t_name, org: org.name, loc: loc.name
+          post :import, params: { "repo" => tmpdir, "filter" => t_name, "prefix" => '', "associate" => "always" }
+          assert_response :success
+          assert_equal ProvisioningTemplate.unscoped.find_by_name(t_name).organizations[0].name, org.name
+          assert_equal ProvisioningTemplate.unscoped.find_by_name(t_name).locations[0].name, loc.name
+        end
+      end
+
+      test "should export filtered template only" do
+        Dir.mktmpdir do |tmpdir|
+          FactoryBot.create(:provisioning_template, :name => 'exporting_template')
+          post :export, params: { "repo" => tmpdir, "filter" => "exporting_template", "metadata_export_mode" => "refresh" }
+          assert_response :success
+          assert Dir.entries("#{tmpdir}/provisioning_templates/provision").include?('exporting_template.erb')
+        end
+      end
+
+      test "should import accordingly to exported template metadata" do
+        org = taxonomies(:organization2)
+        loc = taxonomies(:location2)
+        name = "taxo_metadata_import"
+        templte = "<%#\nmodel: ProvisioningTemplate\nkind: provision\nname: #{name}\norganizations: #{org.name}\nlocations: #{loc.name}\n-%>\nimport metadata"
+        ProvisioningTemplate.create!(:name => name, :template => templte)
```

Comment:
You need to specify `template_kind` for the template to pass the validation.

Reply:
In the template string, yes. But `create!` method does not parse the template string. It just creates a record with given attributes and provisioning template needs template kind association to be [valid](https://github.com/theforeman/foreman/blob/develop/app/models/provisioning_template.rb#L22), therefore you need to have

```ruby
ProvisioningTemplate.create!(:name => name, :template => template, :template_kind => template_kinds(:provision))
```
for validation to pass.

It turns out `ActionController::TestCase` supports only testing [one action](https://api.rubyonrails.org/classes/ActionController/TestCase.html). A workaround might be to call the exporter [directly](https://github.com/theforeman/foreman_templates/blob/master/app/controllers/api/v2/template_controller.rb#L35) or create [integration test](https://api.rubyonrails.org/classes/ActionDispatch/IntegrationTest.html) if you need both requests to happen.